test/e2e/framework.go

```diff
@@ -419,3 +421,172 @@ func kubectlExec(namespace string, podName, containerName string, args ...string
+
+	// Optional: only pods with this name will pass the filter.
+	podName string
+}
+
+type ClusterVerification struct {
+	client    *client.Client
+	namespace *api.Namespace // pointer rather than string, since ns isn't created until before each.
+	podState  PodStateVerification
+}
+
+func (f *Framework) NewClusterVerification(filter PodStateVerification) *ClusterVerification {
+	return &ClusterVerification{
+		f.Client,
+		f.Namespace,
+		filter,
+	}
+}
+
+func passesPodNameFilter(pod api.Pod, name string) bool {
+	return name == "" || strings.Contains(pod.Name, name)
+}
+
+func passesVerifyFilter(pod api.Pod, verify func(p api.Pod) (bool, error)) (bool, error) {
+	if verify == nil {
+		return true, nil
+	} else {
+		verified, err := verify(pod)
+		// If an error is returned, by definition, pod verification fails
+		if err != nil {
+			return false, err
+		} else {
+			return verified, nil
+		}
+
+	}
+}
+
+func passesPhasesFilter(pod api.Pod, validPhases []api.PodPhase) bool {
+	passesPhaseFilter := false
+	for _, phase := range validPhases {
+		if pod.Status.Phase == phase {
+			passesPhaseFilter = true
+		}
+	}
+	return passesPhaseFilter
+}
+
+func filterLabels(selectors map[string]string, cli *client.Client, ns string) (*api.PodList, error) {
+	var err error
+	var selector labels.Selector
+	var pl *api.PodList
+	// List pods based on selectors.  This might be a tiny optimization rather then filtering
+	// everything manually.
+	if len(selectors) > 0 {
+		selector = labels.SelectorFromSet(labels.Set(selectors))
+		options := api.ListOptions{LabelSelector: selector}
+		pl, err = cli.Pods(ns).List(options)
+	} else {
+		pl, err = cli.Pods(ns).List(api.ListOptions{})
+	}
+	return pl, err
+}
+
+// filter pods passing a filter.  It can be used to compose
+// the more useful abstractions like ForEach, WaitFor, and so on, which
+// can be used directly by tests.
+func (p *PodStateVerification) filter(c *client.Client, namespace *api.Namespace) ([]api.Pod, error) {
+	ns := namespace.Name
+	pl, err := filterLabels(p.selectors, c, ns) // Build an api.PodList to operate against.
+	Logf("Selector matched %v pods for %v=%v", len(pl.Items), p.selectors)
+	if len(pl.Items) == 0 || err != nil {
+		return pl.Items, err
+	}
+
+	unfilteredPods := pl.Items
+	filteredPods := []api.Pod{}
+	if len(p.validPhases) == 0 {
+		panic("Need to specify a valid pod phase as part of pod states.")
+	}
+	// Next: Pod must match at least one of the states that the user specified
+	for _, pod := range unfilteredPods {
+		podStatePassesFilter := passesPhasesFilter(pod, p.validPhases)
+
+		podFunctionVerified, err := passesVerifyFilter(pod, p.verify)
+		if err != nil {
+			return filteredPods, err
+		}
+
+		podNameVerified := passesPodNameFilter(pod, p.podName)
+
+		// Compose them all...
+		if podStatePassesFilter && podNameVerified && podFunctionVerified {
+			filteredPods = append(filteredPods, pod)
+		} else {
+			Logf("pod %v state, name, verified %v, %v, %v ",
```

Comment:
per-pod logging will be too much on large tests.  imagine the amount of logging that would be put out on a test cluster of size 1k+ w/ 250 pods/node and checking state.